### PR TITLE
o/patch: check if we have snapd snap with correct snap type already in snapstate

### PR DESCRIPTION
**Pull request by stolowski**
The linter in #10242 uncovered an unused test json which I think was meant to cover particular edge case; it wasn't therefore correctly handled.

The edge case is about `type: snapd` migration in patch 6.2 when having snapd with correct type already in state, but also having an app with snapd's snap-id (but wrong type). I'm not sure if this is a viable scenario, but clearly it was considered when the original change was introduced, but somehow I forgot to wire it up.

### CODE REVIEW

File: overlord/patch/patch6_2.go [hunk 1]
```diff
@@ -112,23 +112,40 @@ func patch6_2(st *state.State) error {
 		return fmt.Errorf("internal error: cannot get snaps: %s", err)
 	}
 
-	// Migrate snapstate
-	for name, raw := range snaps {
+	var hasSnapdSnap bool
```
Comment (mardy) [inline overlord/patch/patch6_2.go:115]: I'm probably going to reveal my golang ignorance now, but don't we need to initialize this to false?

File: overlord/patch/patch6_2.go [hunk 2]
```diff
@@ -112,23 +112,40 @@ func patch6_2(st *state.State) error {
...
+			var snapst patch62SnapState
+			if err := json.Unmarshal([]byte(*raw), &snapst); err != nil {
 				return err
 			}
-			newRaw := json.RawMessage(data)
-			snaps[name] = &newRaw
-			st.Set("snaps", snaps)
-			// We can have at most one snapd snap
-			break
+			if hasSnapdSnapID(snapst) && snapst.SnapType != string(snap.TypeSnapd) {
```
Comment (mardy) [inline overlord/patch/patch6_2.go:137]: isn't the second condition unneeded? if we had a `snapst` for which the expression
```go
    hasSnapdSnapID(snapst) && snapst.SnapType == string(snap.TypeSnapd)
```
held true, then the `hasSnapdSnap` flag would be true, and we wouldn't enter this block.

File: overlord/patch/patch6_2.go (hunk 1 above)
Comment (stolowski) [reply overlord/patch/patch6_2.go:115]: No, go initializes bool to false (and other types to their default values).

File: overlord/patch/patch6_2.go (hunk 2 above)
Comment (stolowski) [reply overlord/patch/patch6_2.go]: Yep, you're right, that can be simplified, thanks!

**Review by mardy — APPROVED** [review]
Looks good, my only concern is that in the worst case we are calling `json.Unmarshal` twice on each snap. I have no idea is this is a problem, though :-)

**Comment by stolowski** [comment]
> Looks good, my only concern is that in the worst case we are calling `json.Unmarshal` twice on each snap. I have no idea if this is a problem, though :-)

Yeah, this is generally a valid concern. In this case this should be fine since this is patch logic and it's run only once (we have patch logic that remembers latest patch level/sublevel and skips already applied patches).

**Review by pedronis — APPROVED** [review]
thanks
